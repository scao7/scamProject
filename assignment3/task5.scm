(define (sdisplay n z)
    (define (iter x m)
        (cond
            ((= m 0) (print "..."))
            (else
                (print (stream-car x) ",")
                (iter (stream-cdr x) (- m 1))
                )
            )
        )

    (print "(")
    (iter z n)
    (print ")")
    )


(define (add-streams s t)
    (cons-stream
        (+ (stream-car s) (stream-car t))
        (add-streams (stream-cdr s) (stream-cdr t))
        )
    )


(define (smush s op)
    (define (trip a b c z)
        (cond
            ((= z 0) (cons-stream (stream-car a) (trip (stream-cdr a) b c 1)))
            ((= z 1) (cons-stream (stream-car b) (trip a (stream-cdr b) c 2)))
            ((= z 2) (cons-stream (stream-car c) (trip a b (stream-cdr c) 0)))
            )
        )


    (define (accum-stream s in)
        (define cur (op in (stream-car s)))
        (cons-stream cur (accum-stream (stream-cdr s) cur))
        )

    (define ones (cons-stream 1 ones))
    (define wholes (cons-stream 0 (add-streams ones wholes)))
    (define temp (cons-stream (stream-car s) (accum-stream (stream-cdr s) (stream-car s))))

    (cons-stream (stream-car wholes) (trip (stream-cdr wholes) s temp 1))
    )


(define (main)
    (define env this)
    (define (iter expr)
         (if (not (eof?)) (begin (eval expr env) (iter (readExpr))))
         )
    (setPort (open (getElement ScamArgs 1) 'read))
    (iter (readExpr))
    )