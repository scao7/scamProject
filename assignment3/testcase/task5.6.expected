op is <builtin +(@)>
(36,31,-76,-33,58,-91,-26,47,85,50,59,59,...)
(0,36,36,1,31,67,2,-76,-9,3,-33,-42,...)
